// grunt-react/backupGrunt.js
/*
 * grunt-react
 * https://github.com/ericclemmons/grunt-react
 *
 * Copyright (c) 2013 Eric Clemmons, contributors
 * Licensed under the MIT license.
 */

'use strict';

module.exports = function(grunt) {
  pkg: grunt.file.readJSON('package.json'),
  // Project configuration.
  grunt.initConfig({
   watch: {
	react: {
	files: '../react_components/*.jsx',
	tasks: ['browserify']
	}
	},
   client: {
	src: ['../react-datagrid/*.jsx'],
	dest: 'scripts/app.js'
	},
    // Used for testing the transformer
    browserify: {
      options: {
        transform: [ require('grunt-react').browserify ]
      },


  });

  // Actually load this plugin's task(s).
 
  // These plugins provide necessary tasks.
  grunt.loadNpmTasks('grunt-browserify');
  grunt.loadNpmTasks('grunt-contrib-watch');
  //grunt.loadNpmTasks('grunt-contrib-clean');
  //grunt.loadNpmTasks('grunt-contrib-nodeunit');

  // Whenever the "test" task is run, first clean the "tmp" dir, then run this
  // plugin's task(s), then test the result.
  //grunt.registerTask('', ['clean', 'react', 'browserify', 'nodeunit']);

  // By default, lint and run all tests.
  grunt.registerTask('default', ['browserify']);

};
