// searchkit/lib/src/components/ui/range/RangeInput.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var React = require("react");
var block = require('bem-cn');
var defaults = require("lodash/defaults");
var get = require("lodash/get");
var clamp = require("lodash/clamp");
/*
 * Input validates input and only calls onChange for valid values
 */
var NumberInput = (function (_super) {
    __extends(NumberInput, _super);
    function NumberInput(props) {
        _super.call(this, props);
        this.onChange = this.onChange.bind(this);
        this.state = {
            value: props.value
        };
    }
    NumberInput.prototype.componentWillReceiveProps = function (nextProps) {
        if (nextProps.value !== this.props.value) {
            this.setState({ value: nextProps.value });
        }
    };
    NumberInput.prototype.isValid = function (value) {
        value = '' + value; // ensure string
        // Weird number check, please do something else
        return ('' + parseInt(value, 10) == value);
    };
    NumberInput.prototype.onChange = function (e) {
        var _a = this.props, field = _a.field, onChange = _a.onChange;
        var value = e.target.value;
        this.setState({ value: value });
        if (this.isValid(value) && onChange) {
            onChange(value, field);
        }
    };
    NumberInput.prototype.render = function () {
        return React.createElement("input", React.__spread({type: "number"}, this.props, {value: this.state.value, onChange: this.onChange}));
    };
    NumberInput.defaultProps = {
        value: ''
    };
    return NumberInput;
}(React.Component));
exports.NumberInput = NumberInput;
var RangeInput = (function (_super) {
    __extends(RangeInput, _super);
    function RangeInput(props) {
        _super.call(this, props);
        // this.handleInputChange = this.handleInputChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }
    RangeInput.prototype.handleInputChange = function (value, key) {
        // const { min, max, minValue, maxValue, onFinished } = this.props
        // const values = defaults({
        //   [key]: clamp(value, min, max)
        // }, {
        //   min: minValue, max: maxValue
        // })
        // onFinished(values)
    };
    RangeInput.prototype.handleSubmit = function (e) {
        e.preventDefault();
        this.props.onFinished({ min: this.refs.min.state.value, max: this.refs.max.state.value });
    };
    RangeInput.prototype.render = function () {
        var _a = this.props, mod = _a.mod, className = _a.className, minValue = _a.minValue, maxValue = _a.maxValue, translate = _a.translate, minPlaceholder = _a.minPlaceholder, maxPlaceholder = _a.maxPlaceholder;
        var bemBlocks = {
            container: block(mod)
        };
        return (React.createElement("form", {className: bemBlocks.container().mix(className), onSubmit: this.handleSubmit}, React.createElement(NumberInput, {ref: "min", className: bemBlocks.container("input"), value: minValue, field: "min", onChange: this.handleInputChange, placeholder: translate('range.min') || minPlaceholder}), React.createElement("div", {className: bemBlocks.container("to-label")}, translate('range.to') || '-'), React.createElement(NumberInput, {ref: "max", className: bemBlocks.container("input"), value: maxValue, field: "max", onChange: this.handleInputChange, placeholder: translate('range.max') || maxPlaceholder}), React.createElement("button", {type: "submit", className: bemBlocks.container("submit")}, translate('range.submit') || 'Go')));
    };
    RangeInput.defaultProps = {
        mod: "sk-range-input",
        translate: function (str) { return undefined; },
        minPlaceholder: 'min',
        maxPlaceholder: 'max',
    };
    return RangeInput;
}(React.Component));
exports.RangeInput = RangeInput;
