// react-notification/dist/notificationStack.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _defaultPropTypes = require('./defaultPropTypes');

var _defaultPropTypes2 = _interopRequireDefault(_defaultPropTypes);

var _stackedNotification = require('./stackedNotification');

var _stackedNotification2 = _interopRequireDefault(_stackedNotification);

/**
 * The notification list does not have any state, so use a
 * pure function here. It just needs to return the stacked array
 * of notification components.
 */
var NotificationStack = function NotificationStack(props) {
  return _react2['default'].createElement(
    'div',
    { className: 'notification-list' },
    props.notifications.map(function (notification, index) {
      var dismissAfter = notification.dismissAfter || props.dismissAfter;
      var lastNotificationDismissAfter = 300;
      var isLast = index === 0 && props.notifications.length === 1;

      return _react2['default'].createElement(_stackedNotification2['default'], _extends({}, notification, {
        key: notification.key,
        isLast: isLast,
        action: notification.action || props.action,
        dismissAfter: isLast ? 2000 : 2000 + index * 1000,
        onClick: function () {
          return props.onDismiss(notification);
        },
        onDismiss: function () {
          return setTimeout(function () {
            setTimeout(props.onDismiss.bind(undefined, notification), lastNotificationDismissAfter);
          }, dismissAfter !== null ? dismissAfter : 2000);
        },
        style: {
          bar: {
            bottom: 2 + index * 4 + 'rem'
          }
        }
      }));
    })
  );
};

NotificationStack.propTypes = {
  notifications: _react.PropTypes.array.isRequired,
  onDismiss: _react.PropTypes.func
};

exports['default'] = NotificationStack;
module.exports = exports['default'];
