// react-inspector/lib/object/ObjectPreview.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _ObjectDescription = require('./ObjectDescription');

var _ObjectDescription2 = _interopRequireDefault(_ObjectDescription);

var _ObjectName = require('./ObjectName');

var _ObjectName2 = _interopRequireDefault(_ObjectName);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var styles = {
  preview: {
    fontStyle: 'italic'
  }
};

function intersperse(arr, sep) {
  if (arr.length === 0) {
    return [];
  }

  return arr.slice(1).reduce(function (xs, x, i) {
    return xs.concat([sep, x]);
  }, [arr[0]]);
}

/**
 * A preview of the object
 * if a name is specified, it will render a simplified preview with a short description
 */
var ObjectPreview = function ObjectPreview(_ref) {
  var maxProperties = _ref.maxProperties;
  var object = _ref.object;
  var name = _ref.name;

  if (typeof name !== 'undefined') {

    var Colon = function Colon() {
      return _react2.default.createElement(
        'span',
        null,
        ': '
      );
    };
    return _react2.default.createElement(
      'span',
      null,
      _react2.default.createElement(_ObjectName2.default, { name: name }),
      _react2.default.createElement(Colon, null),
      _react2.default.createElement(_ObjectDescription2.default, { object: object })
    );
  }

  if ((typeof object === 'undefined' ? 'undefined' : _typeof(object)) !== 'object' || object === null) {
    return _react2.default.createElement(_ObjectDescription2.default, { object: object });
  }

  if (Array.isArray(object)) {
    return _react2.default.createElement(
      'span',
      { style: styles.preview },
      '[',
      intersperse(object.map(function (element, index) {
        return _react2.default.createElement(_ObjectDescription2.default, { key: index, object: element });
      }), ", "),
      ']'
    );
  } else if (object instanceof Date) {
    return _react2.default.createElement(
      'span',
      null,
      object.toString()
    );
  } else {
    var propertyNodes = [];
    for (var propertyName in object) {
      var propertyValue = object[propertyName];
      if (object.hasOwnProperty(propertyName)) {
        var ellipsis = undefined;
        if (propertyNodes.length === maxProperties - 1 && Object.keys(object).length > maxProperties) {
          ellipsis = _react2.default.createElement(
            'span',
            { key: 'ellipsis' },
            '…'
          );
        }
        propertyNodes.push(_react2.default.createElement(
          'span',
          { key: propertyName },
          _react2.default.createElement(_ObjectName2.default, { name: propertyName }),
          ': ',
          _react2.default.createElement(_ObjectDescription2.default, { object: propertyValue }),
          ellipsis
        ));
        if (ellipsis) break;
      }
    }

    return _react2.default.createElement(
      'span',
      { style: styles.preview },
      'Object {',
      intersperse(propertyNodes, ", "),
      '}'
    );
  }
};

ObjectPreview.propTypes = {
  maxProperties: _react2.default.PropTypes.number
};
ObjectPreview.defaultProps = {
  maxProperties: 5 /* max number of properties shown in the property view */
};

exports.default = ObjectPreview;
